### ISSUE TITLE
FP single_component_path_imports

### ISSUE THREAD

**Issue by djc** [rust-lang/rust-clippy#7168]
Lint name:

single_component_path_imports

I tried this code:

```rust
#[cfg(feature = "log")]
use log;

mod foo {
    use crate::log::warn;

    pub fn bar() {
        warn!("bar");
    }
}

pub fn bar() {
    foo::bar()
}
```

(I think it doesn't matter whether the feature `log` is turned on or off.)

I expected to see this happen: no warning, since the `log` import is used in an inner module.

Instead, this happened:

```rust
warning: this import is redundant
 --> src/lib.rs:3:1
  |
3 | use log;
  | ^^^^^^^^ help: remove it entirely
  |
  = note: `#[warn(clippy::single_component_path_imports)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_component_path_imports
```

### Meta

- `cargo clippy -V`: clippy 0.1.53 (42816d61 2021-04-24)
- `rustc -Vv`:
  ```
  rustc 1.53.0-nightly (42816d61e 2021-04-24)
  binary: rustc
  commit-hash: 42816d61ead7e46d462df997958ccfd514f8c21c
  commit-date: 2021-04-24
  host: x86_64-apple-darwin
  release: 1.53.0-nightly
  LLVM version: 12.0.0
  ```

**Comment by camsteffen** [comment]
So you want the `log` crate to be accessible at `crate::log`?

**Comment by djc** [comment]
It is accessible as `crate::log`, and my code relies on that.

**Comment by camsteffen** [comment]
Yes, obviously the suggestion is not okay as is. I'm just asking from a code style viewpoint. Because this questions the assertion that "single component imports are always useless/bad style" which seems to be the idea behind the lint.

**Comment by djc** [comment]
From a code style viewpoint I think it's a great lint, I've applied many suggestions due to this lint. In this case the `use log` is conditional on a feature and gets switched out for a mock implementation of the log warnings, so it's definitely more of a special case. In general if the identifier gets used as a crate-rooted item, or maybe even in general through a path, the lint should probably be suppressed.

**Comment by camsteffen** [comment]
> In general if the identifier gets used as a crate-rooted item, or maybe even in general through a path, the lint should probably be suppressed.

Yep that sounds reasonable. Hopefully it isn't too gnarly to implement. Scanning usages can be problematic. But I think this might be a relatively simple case.

**Comment by camsteffen** [comment]
We should also detect usages like `super::log` which means it doesn't need to be crate-rooted.
